File: tests/machine/x/dart/group_by_multi_join.dart
var nations = [
  {'id': 1, 'name': 'A'},
  {'id': 2, 'name': 'B'},
];

var suppliers = [
  {'id': 1, 'nation': 1},
  {'id': 2, 'nation': 2},
];

var partsupp = [
  {'part': 100, 'supplier': 1, 'cost': 10, 'qty': 2},
  {'part': 100, 'supplier': 2, 'cost': 20, 'qty': 1},
  {'part': 200, 'supplier': 1, 'cost': 5, 'qty': 3},
];

var filtered = (() {
  var _q0 = <dynamic>[];
  for (var ps in partsupp) {
    for (var s in suppliers) {
      if (!(s['id'] == ps['supplier'])) continue;
      for (var n in nations) {
        if (!(n['id'] == s['nation'])) continue;
        if (!(n['name'] == 'A')) continue;
        _q0.add({
          'part': ps['part'],
          'value': (ps['cost'] as num) * (ps['qty'] as num),
        });
      }
    }
  }
  return _q0;
})();

var grouped = (() {
  var _q1 = <dynamic>[];
  var _g2 = <dynamic, List<dynamic>>{};
  for (var x in filtered) {
    var _k4 = x['part'];
    _g2.putIfAbsent(_k4, () => <dynamic>[]).add({'x': x});
  }
  for (var entry in _g2.entries) {
    var g = entry.value;
    var _k4 = entry.key;
    _q1.add({
      'part': _k4,
      'total': (() {
        var _q5 = <dynamic>[];
        for (var r in g) {
          _q5.add(r['value']);
        }
        return _q5;
      })().reduce((a, b) => a + b),
    });
  }
  return _q1;
})();

void main() {
  print(grouped);
}

File: tests/machine/x/rust/group_by_multi_join.rs
#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
struct Nation {
    id: i32,
    name: &'static str,
}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
struct Supplier {
    id: i32,
    nation: i32,
}

#[derive(Default, Debug, Clone, PartialEq)]
struct Partsupp {
    part: i32,
    supplier: i32,
    cost: f64,
    qty: i32,
}

#[derive(Default, Debug, Clone, PartialEq)]
struct Result {
    part: i32,
    value: f64,
}

#[derive(Default, Debug, Clone, PartialEq)]
struct Group {
    key: i32,
    items: Vec<Result>,
}

#[derive(Default, Debug, Clone, PartialEq)]
struct Result1 {
    part: i32,
    total: f64,
}

fn sum<T>(v: &[T]) -> T where T: std::iter::Sum<T> + Copy {
    v.iter().copied().sum()
}

fn main() {
    let nations = vec![Nation { id: 1, name: "A" }, Nation { id: 2, name: "B" }];
    let suppliers = vec![Supplier { id: 1, nation: 1 }, Supplier { id: 2, nation: 2 }];
    let partsupp = vec![Partsupp { part: 100, supplier: 1, cost: 10.0, qty: 2 }, Partsupp { part: 100, supplier: 2, cost: 20.0, qty: 1 }, Partsupp { part: 200, supplier: 1, cost: 5.0, qty: 3 }];
    let filtered = { let mut tmp1 = Vec::new();for ps in &partsupp { for s in &suppliers { if !(s.id == ps.supplier) { continue; } for n in &nations { if !(n.id == s.nation) { continue; } if !(n.name == "A") { continue; } tmp1.push(Result { part: ps.part, value: ps.cost * ps.qty as f64 }); } } } tmp1 };
    let grouped = { let mut tmp2 = std::collections::HashMap::new();for x in &filtered { let key = x.part; tmp2.entry(key).or_insert_with(Vec::new).push(x.clone()); } let mut tmp3 = Vec::<Group>::new(); for (k,v) in tmp2 { tmp3.push(Group { key: k, items: v }); } let mut result = Vec::new(); for g in tmp3 { result.push(Result1 { part: g.key, total: sum(&{ let mut tmp4 = Vec::new();for r in &g.clone().items { tmp4.push(r.value); } tmp4 }) }); } result };
    println!("{:?}", grouped);
}
